play-json/src/main/scala/play/api/libs/json/JsMacroImpl.scala

```diff
@@ -263,236 +371,286 @@ object JsMacroImpl {
 
-      private def createImplicit(name: Name, ptype: Type): Implicit = {
-        val tpe = ptype match {
-          case TypeRef(_, _, targ :: _) =>
-            // Option[_] needs special treatment because we need to use XXXOpt
-            if (ptype.typeConstructor <:< typeOf[Option[_]].typeConstructor) {
-              targ
-            } else ptype
+    def macroSealedFamilyImpl(subTypes: List[Type]): c.Expr[M[A]] = {
+      val typeNaming = (_: Type).typeSymbol.fullName
 
-          case SingleType(_, _) | TypeRef(_, _, _) => ptype
-        }
+      def readLambda: Tree = {
+        val resolver = new ImplicitResolver({ orig: Type => orig })
+        val cases = Match(q"""obj.value.lift("_value")""", (
+          cq"""_ => $json.JsError("error.missing.path")""" :: (
+            subTypes.foldLeft(List.empty[CaseDef]) { (out, t) =>
+              val rtpe = appliedType(readsType, List(t))
+              val reader = resolver.createImplicit(
+                atag.tpe, rtpe
+              )(t).neededImplicit
 
-        val (ntpe, selfRef) = normalized(tpe)
-        val ptpe = boundTypes.getOrElse(ntpe.typeSymbol.fullName, ntpe)
+              if (reader.isEmpty) {
+                c.abort(
+                  c.enclosingPosition,
+                  s"No instance of Reads is available for ${t.typeSymbol.fullName} in the implicit scope (Hint: if declared in the same file, make sure it's declared before)"
+                )
+              }
 
-        // infers implicit
-        val neededImplicitType = appliedType(natag.tpe.typeConstructor, ptpe)
-        val neededImplicit = if (!selfRef) {
-          c.inferImplicitValue(neededImplicitType)
-        } else c.untypecheck(
-          // Reset the type attributes on the refactored tree for the implicit
-          ImplicitTransformer.transform(
-            c.inferImplicitValue(neededImplicitType)
+              cq"""Some(v) if (dis == ${typeNaming(t)}) => $reader.reads(v)""" :: out
+            }
           )
-        )
-
-        Implicit(name, ptype, neededImplicit, tpe, selfRef)
-      }
-
-      // To print the implicit types in the compiler messages
-      private def prettyType(t: Type): String =
-        boundTypes.getOrElse(t.typeSymbol.fullName, t) match {
-          case TypeRef(_, base, args) if args.nonEmpty => s"""${base.asType.fullName}[${args.map(prettyType).mkString(", ")}]"""
-
-          case tp => tp.typeSymbol.fullName
-        }
+        ).reverse)
 
-      def apply(params: List[Symbol]): List[Implicit] = {
-        val resolvedImplicits = params.map { param =>
-          createImplicit(param.name, param.typeSignature)
-        }
-        val effectiveImplicits = if (ApplyUnapply.hasVarArgs) {
-          val varArgsImplicit = createImplicit(
-            resolvedImplicits.last.paramName,
-            ApplyUnapply.unapplyReturnTypes.get.last
-          )
+        q"""(_: $json.JsValue) match {
+          case obj @ $json.JsObject(_) => obj.value.get("_type") match {
```

Comment:
Using `JsonConfiguration` is planned to customize the naming (name of the discriminator field, value for this field accord each case/type).

I personnally feel it could be done in a separate PR, to validate/merge step by step, as long as each step is validated and backward compatible.